lib/canister.rb

```diff
@@ -6,18 +6,49 @@
 # resolved at runtime. This allows for out-of-order declaration,
 # automatic dependency resolution, and--upon
 # redeclaration--automatic dependency cache invalidation.
+# @param hashlike [#each_pair, #keys #[]]
+# @see #fill_from_hashlike
 class Canister
-  def initialize
+  def initialize(hashlike = nil)
     @stack = []
     @registry = {}
     @resolved = {}
     @dependents = Hash.new do |hash, key|
       hash[key] = []
     end
     @mutex = Mutex.new
+    if hashlike
+      self.merge_from_hashlike(hashlike)
+    end
     yield self if block_given?
   end
 
+  # A "hashlike" is defined as anything that responds to both #[]
+  # and #keys and/or #each_pair
+  # Each value is wrapped in a `proc` and registered under the key.
+  # Note that Canister doesn't differentiate between symbols and
+  # strings for keys, so if your hashlike has keys of, e.g. both
+  # `"a"` and `:a` it won't work.
```

Comment:
I would recommend expanding on this and making explicit the fact that keys symbolizing to the same value will overwrite one another in unpredictable ways. It's hard to imagine a case in which this would be the expected or in any way legit usage so I would even put something like this in `merge_from_hashlike`

```
if hashlike.keys.map(&:to_sym).uniq.count != hashlike.count
  raise "you're gonna lose info!"
end
```